File: comms/server/controllers/mywcom_loc_controller.py
from pyramid.view import view_config
import json

from myworldapp.core.server.base.db.globals import Session
from myworldapp.modules.comms.server.api.network_view import NetworkView

from myworldapp.core.server.controllers.base.myw_feature_collection import (
    MywFeatureCollection,
)
from myworldapp.modules.comms.server.controllers.mywcom_controller import (
    MywcomController,
)


class MywcomLocController(MywcomController):
    """
    Tracing logic in this class is based on the NM show terminations code
    """

    def __init__(self, request):
        """
        Initialize slots of self
        """

        super().__init__(request, "LOGICAL COUNTS")

    @view_config(
        route_name="mywcom_loc_controller.ripple_trace",
        request_method="GET",
        renderer="json",
    )
    def ripple_trace(self):
        """
        Performs a ripple trace. Basically, a normal pin level trace with connectivity information returned
        that will then be used to update line of count information
        """

        feature_type = self.get_param(self.request, "feature_type", mandatory=True)
        id = self.get_param(self.request, "id", mandatory=True)
        do_update = self.get_param(self.request, "update", default=False)

        app = self.get_param(self.request, "application")
        delta = self.get_param(self.request, "delta")
        side = self.get_param(self.request, "side")
        config = self.get_param(self.request, "config")
        config = json.loads(config) if config else None

        # Check authorised
        self.current_user.assertAuthorized(self.request, feature_type=feature_type, application=app)

        db_view = self.db.view(delta)
        nw_view = NetworkView(db_view)
        loc_mgr = nw_view.loc_mgr

        result = loc_mgr.rippleTrace(feature_type, id, side=side, config=config)

        # Update loc segment records along the trace if asked to.
        if do_update:
            feature = db_view.get(f"{feature_type}/{id}")
            loc_mgr.rippleUpdate(result, feature, side=side)

        return result

    @view_config(
        route_name="mywcom_loc_controller.ripple_trace_update",
        request_method="POST",
        renderer="json",
    )
    def ripple_trace_update(self):
        """
        Performs a ripple trace and then update.
        """

        feature_type = self.get_param(self.request, "feature_type", mandatory=True)
        id = self.get_param(self.request, "id", mandatory=True)

        app = self.get_param(self.request, "application")
        delta = self.get_param(self.request, "delta")
        side = self.get_param(self.request, "side")

        # Check authorised
        self.current_user.assertAuthorized(
            self.request, right="editFeatures", feature_type="mywcom_line_of_count", application=app
        )

        db_view = self.db.view(delta)
        nw_view = NetworkView(db_view)
        loc_mgr = nw_view.loc_mgr

        result = loc_mgr.rippleTrace(feature_type, id, side=side)

        feature = db_view.get(f"{feature_type}/{id}")
        loc_mgr.rippleUpdate(result, feature, side=side)

        Session.commit()

        return result

    @view_config(
        route_name="mywcom_loc_controller.get_loc",
        request_method="POST",
        renderer="json",
    )
    def get_loc(self):
        """
        Fetch line of count information for multiple features
        """

        app = self.get_param(self.request, "application")
        delta = self.get_param(self.request, "delta")
        feature_qurns = json.loads(self.get_param(self.request, "urns", mandatory=True))
        include_proposed = self.get_param(
            self.request, "include_proposed", default=False, type=bool
        )

        # Check authorised
        self.current_user.assertAuthorized(
            self.request, feature_type="mywcom_line_of_count", application=app
        )

        db_view = self.db.view(delta)
        nw_view = NetworkView(db_view)
        loc_mgr = nw_view.loc_mgr

        feature_loc = loc_mgr.getLocMany(feature_qurns, include_proposed=include_proposed)

        return feature_loc

    @view_config(
        route_name="mywcom_loc_controller.get_loc_details",
        request_method="POST",
        renderer="json",
    )
    def get_loc_details(self):
        """
        Fetch line of count information for multiple features
        """

        app = self.get_param(self.request, "application")
        delta = self.get_param(self.request, "delta")
        feature_qurns = json.loads(self.get_param(self.request, "urns", mandatory=True))
        include_proposed = self.get_param(
            self.request, "include_proposed", default=False, type=bool
        )

        # Check authorised
        self.current_user.assertAuthorized(
            self.request, feature_type="mywcom_line_of_count", application=app
        )

        db_view = self.db.view(delta)
        nw_view = NetworkView(db_view)
        loc_mgr = nw_view.loc_mgr

        loc_data = loc_mgr.getLocDetailsMany(feature_qurns, include_proposed=include_proposed)

        return loc_data

    @view_config(
        route_name="mywcom_loc_controller.update_loc",
        request_method="POST",
        renderer="json",
    )
    def update_loc(self):
        """
        Updates line of count information for multiple features
        """

        app = self.get_param(self.request, "application")
        delta = self.get_param(self.request, "delta")
        feature_loc = self.get_param(self.request, "feature_loc")
        mark_stale = self.get_param(self.request, "mark_stale", default=False, type=bool)

        # Check authorised
        self.current_user.assertAuthorized(
            self.request, right="editFeatures", feature_type="mywcom_line_of_count", application=app
        )

        db_view = self.db.view(delta)
        nw_view = NetworkView(db_view)
        loc_mgr = nw_view.loc_mgr

        feature_loc = json.loads(feature_loc)

        loc_mgr.updateLocMany(feature_loc,mark_stale)

        feature_qurns = feature_loc.keys()

        Session.commit()

        loc_data = loc_mgr.getLocMany(feature_qurns)

        return loc_data

    @view_config(
        route_name="mywcom_loc_controller.ripple_deletions",
        request_method="POST",
        renderer="json",
    )
    def ripple_deletions(self):

        feature_type = self.get_param(self.request, "feature_type", mandatory=True)
        id = self.get_param(self.request, "id", mandatory=True)

        app = self.get_param(self.request, "application")
        delta = self.get_param(self.request, "delta")
        data = self.get_param(self.request, "loc_data")
        side = self.get_param(self.request, "side")

        # Check authorised
        self.current_user.assertAuthorized(
            self.request, right="editFeatures", feature_type=feature_type, application=app
        )

        db_view = self.db.view(delta)
        nw_view = NetworkView(db_view)
        loc_mgr = nw_view.loc_mgr

        feature = db_view.get(f"{feature_type}/{id}")

        updates = loc_mgr.rippleDeletions(feature, side=side)

        Session.commit()

        return {"update": updates}

    @view_config(
        route_name="mywcom_loc_controller.disconnect_loc",
        request_method="POST",
        renderer="json",
    )
    def disconnect_loc(self):
        """
        Updates line of count information after a disconnect
        """

        feature_type = self.get_param(self.request, "feature_type", mandatory=True)
        id = self.get_param(self.request, "id", mandatory=True)
        side = self.get_param(self.request, "side")
        ripple = self.get_param(self.request, "ripple", default=False, type=bool)

        app = self.get_param(self.request, "application")
        delta = self.get_param(self.request, "delta")

        # Check authorised
        self.current_user.assertAuthorized(
            self.request, right="editFeatures", feature_type="mywcom_line_of_count", application=app
        )

        db_view = self.db.view(delta)
        nw_view = NetworkView(db_view)
        loc_mgr = nw_view.loc_mgr

        feature = db_view.get(f"{feature_type}/{id}")

        result = loc_mgr.disconnectLoc(feature, side=side, ripple=ripple)

        Session.commit()

        return result

    @view_config(
        route_name="mywcom_loc_controller.connect_loc",
        request_method="POST",
        renderer="json",
    )
    def connect_loc(self):
        """
        Updates line of count information after a connection. Connection record is passed in.
        """

        feature_type = self.get_param(self.request, "feature_type", mandatory=True)
        id = self.get_param(self.request, "id", mandatory=True)
        ripple = self.get_param(self.request, "ripple", default=False, type=bool)

        app = self.get_param(self.request, "application")
        delta = self.get_param(self.request, "delta")

        # Check authorised
        self.current_user.assertAuthorized(
            self.request, right="editFeatures", feature_type=feature_type, application=app
        )

        db_view = self.db.view(delta)
        nw_view = NetworkView(db_view)
        loc_mgr = nw_view.loc_mgr

        # Get connection record
        feature = db_view.get(f"{feature_type}/{id}")

        result = loc_mgr.connectLoc(feature, ripple=ripple)

        Session.commit()

        return result

File: comms/native/services/controllers/MywcomLocController.js
// import view_config from 'pyramid.view';

// import { Session, MywFeatureCollection } from 'myWorld-native-services';
import LOCManager from '../api/LocManager';
import NetworkView from '../api/NetworkView';
import MywcomController from './MywcomController';

/**
 * Tracing logic in this class is based on the NM show terminations code
 */
class MywcomLocController extends MywcomController {
    async ripple_trace(routeParams, params) {
        return this.runInTransaction(async () => this._ripple_trace(routeParams, params));
    }

    async _ripple_trace(routeParams, params) {
        const feature_type = routeParams.feature_type;
        const id = routeParams.id;

        const do_update = params.do_update;
        const delta = params.delta;
        const side = params.side;
        const config = JSON.parse(params.config);

        const db_view = this.db.view(delta);
        const nw_view = new NetworkView(db_view);
        const loc_mgr = new LOCManager(nw_view);

        const result = await loc_mgr.rippleTrace(feature_type, id, side, config);

        // Update loc segment records along the trace if requested
        if (do_update) {
            const feature = await db_view.get(`${feature_type}/${id}`);
            await loc_mgr.rippleUpdate(result, feature, side);
        }

        return result;
    }

    async ripple_trace_update(routeParams, params) {
        return this.runInTransaction(async () => this._ripple_trace_update(routeParams, params));
    }

    async _ripple_trace_update(routeParams, params) {
        const feature_type = routeParams.feature_type;
        const id = routeParams.id;

        const delta = params.delta;
        const side = params.side;

        const db_view = this.db.view(delta);
        const nw_view = new NetworkView(db_view);
        const loc_mgr = new LOCManager(nw_view);

        const result = await loc_mgr.rippleTrace(feature_type, id, side);

        const feature = await db_view.get(`${feature_type}/${id}`);
        await loc_mgr.rippleUpdate(result, feature, side);

        return result;
    }

    async get_loc(routeParams, params) {
        const delta = params.delta;
        const feature_qurns = JSON.parse(params.urns);
        const include_proposed = params.include_proposed;

        const db_view = this.db.view(delta);
        const nw_view = new NetworkView(db_view);
        const loc_mgr = new LOCManager(nw_view);

        const loc_data = await loc_mgr.getLocMany(feature_qurns, { include_proposed });

        return loc_data;
    }

    async get_loc_details(routeParams, params) {
        const delta = params.delta;
        const feature_qurns = JSON.parse(params.urns);
        const include_proposed = params.include_proposed;

        const db_view = this.db.view(delta);
        const nw_view = new NetworkView(db_view);
        const loc_mgr = new LOCManager(nw_view);

        const loc_data = await loc_mgr.getLocDetailsMany(feature_qurns, include_proposed);

        return loc_data;
    }

    async update_loc(routeParams, params) {
        return this.runInTransaction(async () => this._update_loc(routeParams, params));
    }

    async _update_loc(routeParams, params) {
        const delta = params.delta;
        const feature_loc_json = params.feature_loc;

        const db_view = this.db.view(delta);
        const nw_view = new NetworkView(db_view);
        const loc_mgr = new LOCManager(nw_view);

        const feature_loc = JSON.parse(feature_loc_json);
        const mark_stale = params.mark_stale;

        await loc_mgr.updateLocMany(feature_loc, mark_stale);

        const feature_qurns = Object.keys(feature_loc);

        const update_loc_data = await loc_mgr.getLocMany(feature_qurns);

        return update_loc_data;
    }

    async ripple_deletions(routeParams, params) {
        return this.runInTransaction(async () => this._ripple_deletions(routeParams, params));
    }

    async _ripple_deletions(routeParams, params) {
        const feature_type = routeParams.feature_type;
        const id = routeParams.id;

        const delta = params.delta;
        const side = params.side;

        const db_view = this.db.view(delta);
        const nw_view = new NetworkView(db_view);
        const loc_mgr = new LOCManager(nw_view);

        const feature = await db_view.get(`${feature_type}/${id}`);

        const updates = await loc_mgr.rippleDeletions(feature, side);

        return { update: updates };
    }

    async disconnect_loc(routeParams, params) {
        return this.runInTransaction(async () => this._disconnect_loc(routeParams, params));
    }

    async _disconnect_loc(routeParams, params) {
        const feature_type = routeParams.feature_type;
        const id = routeParams.id;

        const side = params.side;
        const ripple = params.ripple;
        const delta = params.delta;

        const db_view = this.db.view(delta);
        const nw_view = new NetworkView(db_view);
        const loc_mgr = new LOCManager(nw_view);

        const feature = await db_view.get(`${feature_type}/${id}`);

        const result = await loc_mgr.disconnectLoc(feature, side, ripple);

        return result;
    }

    async connect_loc(routeParams, params) {
        return this.runInTransaction(async () => this._connect_loc(routeParams, params));
    }

    async _connect_loc(routeParams, params) {
        const feature_type = routeParams.feature_type;
        const id = routeParams.id;

        const side = params.side;
        const ripple = params.ripple;
        const delta = params.delta;

        const db_view = this.db.view(delta);
        const nw_view = new NetworkView(db_view);
        const loc_mgr = new LOCManager(nw_view);

        const feature = await db_view.get(`${feature_type}/${id}`);

        const result = await loc_mgr.connectLoc(feature, ripple);

        return result;
    }
}

export default MywcomLocController;
